tensorboard/plugins/graph/keras_util.py

```diff
@@ -92,89 +121,171 @@ def _is_model(layer):
   return layer.get('config').get('layers') is not None
 
 
-def _norm_to_list_of_layers(maybe_layers):
-  """Normalizes to a list of layers.
+def _get_inbound_nodes(tensor_key_with_args_array):
+  """Return normalized list of TensorKeyWithArgs.
+
+  Keras model serialization records input to a layer in a property,
+  `tensor_key_with_args_array`, and it can be of below forms:
+  - [['name_1', 1, 0], ['name_2', 1, 1]]: two invocations witn single input.
+  - [
+      [['name_1', 0, 0], ['name_2', 1, 0]],
+      [['name_3', 0, 0], ['name_4', 1, 0]]]: two invocations with multiple
+    inputs. Multiple inputs are only applicable to limited keras layers like
+    keras.layers.add, keras.layers.concat, and etc...
+
+  Note that there cannot be fixture of types as a layer that takes multiple
+  inputs have precondition on size of the input.
 
   Args:
-    maybe_layers: A list of data[1] or a list of list of data.
+    tensor_key_with_args_array: Value of `tensor_key_with_args_array` of a Keras
+    model serialization.
 
-  Returns:
-    List of list of data.
+  Raises:
+    ValueError: when the `tensor_key_with_args_array` violate our assumption
+        that a layer not being able to take multiple forms of input.
 
-  [1]: A Functional model has fields 'inbound_nodes' and 'output_layers' which can
-  look like below:
+  Returns:
+    List of invocations that has a list of inputs.
+  """
+  normalized_key_array_list = tensor_key_with_args_array \
+      if tensor_key_with_args_array else []
+
+  is_single_input_mode = isinstance(
+      normalized_key_array_list[0][0], six.string_types) \
+          if normalized_key_array_list else False
+  invocations = []
+  for key_array in normalized_key_array_list:
+    if (is_single_input_mode and not
+        isinstance(key_array[0], six.string_types)):
+      raise ValueError('Expected all inbound nodes to have the same type')
+
+    if is_single_input_mode:
+      invocations.append([key_array])
+    else:
+      invocations.append(key_array)
+
+  normalized_key_array_list = []
+  for invocation in invocations:
+    inbound_invocation = []
+    for input_layer in invocation:
+      inbound_invocation.append(
+          TensorKeyWithArgs(
+              inbound_layer=input_layer[0],
+              node_index=input_layer[1],
+              tensor_index=input_layer[2],
+              node_args=input_layer[3]))
+    normalized_key_array_list.append(inbound_invocation)
+  return normalized_key_array_list
+
+
+def _get_tensor_key(tensor_key_array):
+  """Return normalized list of TensorKeyArray and returns list of TensorKeys.
+
+  A model configuration has fields 'input_layers' and 'output_layers' which can
+  be of below forms:
```

Comment:
If I understand correctly, these two forms correspond to “just one
input” and “multiple inputs” respectively. In particular, in the first
case, the length is a constant (3), where in the second case the length
is variable but the length of each element is the same constant (3). Is
that right?

If so, then could you please note this? You’ve defined enough vocabulary
to express it succinctly:

```python
  """...

  A model configuration has fields `input_layers` and `output_layers`.
  Each is either a raw `TensorKey` or a list of raw `TensorKey`s. This
  function normalizes such a value to always be a list of `TensorKey`s,
  forming a singleton list in the case where the input is a single raw
  `TensorKey`.

  ...

  Returns:
    A list of `TensorKey` values.
  """
```